lxd/api_cluster.go

```diff
@@ -2233,3 +2241,531 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
 
 	return response.SyncResponse(true, nil)
 }
+
+// swagger:operation POST /1.0/cluster/members/{name}/state cluster cluster_member_state_post
+//
+// Evacuate or restore a cluster member
+//
+// Evacuates or restores a cluster member.
+//
+// ---
+// consumes:
+//   - application/json
+// produces:
+//   - application/json
+// parameters:
+//   - in: body
+//     name: cluster
+//     description: Cluster member state
+//     required: true
+//     schema:
+//       $ref: "#/definitions/ClusterMemberStatePost"
+// responses:
+//   "200":
+//     $ref: "#/responses/EmptySyncResponse"
+//   "400":
+//     $ref: "#/responses/BadRequest"
+//   "403":
+//     $ref: "#/responses/Forbidden"
+//   "500":
+//     $ref: "#/responses/InternalServerError"
+func clusterNodeStatePost(d *Daemon, r *http.Request) response.Response {
```

Comment:
I would have expected to see some logic here to check `mux.Vars(r)["name"]` against the local name and if it doesn't match, forward the request to the correct server.

As it stands, I'm not seeing forwarding logic, so I suspect that if I hit `/1.0/cluster/members/vm03` on `vm01`, I'll end up with `vm01` getting evacuated rather than `vm03`.

Reply:
The correct cluster member is evacuated or restored in the functions `evacuateClusterMember` and `restoreClusterMember` respectively, without having to forward the request. The test confirms this as we hit `/1.0/cluster/members/node1` on `node2`.

Reply:
Ah yeah, I see, you're using the API to shutdown the remote instances. This works but is going to be really slow.

It'd be far preferable to just forward the state change request directly to the correct server which can then load the instances and handle the shutdown operation without any overhead, only establishing remote connections for the migration stage.